### PR TITLE
Test model outputs equivalence

### PR DESCRIPTION
**Pull request by LysandreJik**
Adds a test to check that the model outputs keep the same values and order as the tuple output.

### CODE REVIEW

File: src/transformers/modeling_tf_longformer.py [hunk 1]
```diff
@@ -992,7 +1004,12 @@ def call(
 
         embedding_output = self.embeddings(input_ids, position_ids, token_type_ids, inputs_embeds, training=training)
         encoder_outputs = self.encoder(
-            [embedding_output, extended_attention_mask, output_attentions, output_hidden_states, padding_len],
+            embedding_output,
+            attention_mask=extended_attention_mask,
+            padding_len=padding_len,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_longformer.py:1012]: @patrickvonplaten After the previous fix was done, there was an issue about `output_attentions` being a `tf.Tensor` that couldn't be used as a Python bool. The fix was to pass the arguments to the encoder as kwargs instead of as a dict.

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -63,14 +68,46 @@ class ModelTesterMixin:
...
+            inputs_dict = {
                 k: v.unsqueeze(1).expand(-1, self.model_tester.num_choices, -1).contiguous()
                 if isinstance(v, torch.Tensor) and v.ndim > 1
                 else v
                 for k, v in inputs_dict.items()
             }
+
+        if return_labels:
+            if model_class in MODEL_FOR_MULTIPLE_CHOICE_MAPPING.values():
+                inputs_dict["labels"] = torch.ones(self.model_tester.batch_size, dtype=torch.long, device=torch_device)
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:83]: Forgot to add this in torch when I added the loss computation tests in TF. Will add tests for loss computation in torch soon.

File: tests/test_modeling_t5.py [hunk 3]
```diff
@@ -37,6 +37,8 @@ def __init__(self, parent):
         self.batch_size = 13
         self.encoder_seq_length = 7
         self.decoder_seq_length = 9
+        # For common tests
+        self.seq_length = self.decoder_seq_length
```
Comment (LysandreJik) [inline tests/test_modeling_t5.py:41]: @patrickvonplaten this value is needed for the labels building. Is it alright if I add this?

File: src/transformers/modeling_tf_longformer.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_tf_longformer.py:1012]: I see, that's were good to know, thanks!

**Comment by patrickvonplaten** [comment]
Awesome that we can remove the `cast_to_bool` hack here. Maybe we can remove it in `t5_modeling_tf_` as well

File: tests/test_modeling_common.py (hunk 2 above)
Comment (patrickvonplaten) [reply tests/test_modeling_common.py:83]: (nit) The last 4 elifs are identical maybe we can shorten the code there a bit with one long if statement

**Review by patrickvonplaten — APPROVED** [review]
Awesome. Great test to check for output equivalence!

File: src/transformers/modeling_tf_longformer.py [hunk 4]
```diff
@@ -833,33 +837,41 @@ def __init__(self, config, **kwargs):
...
-        if cast_bool_to_primitive(output_hidden_states, self.output_hidden_states) is True:
-            outputs = outputs + (all_hidden_states,)
-        if cast_bool_to_primitive(output_attentions, self.output_attentions) is True:
-            outputs = outputs + (all_attentions,)
-        return outputs  # outputs, (hidden states), (attentions)
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states, all_attentions] if v is not None)
+        return TFBaseModelOutput(
+            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_attentions
+        )
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/modeling_tf_longformer.py:874]: Good catch!

File: src/transformers/modeling_tf_longformer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_tf_longformer.py:1012]: It would be good to try the same in TF T5 which also passes the inputs as a a big list. Maybe this was the cause of our bugs.

File: tests/test_modeling_tf_common.py [hunk 5]
```diff
@@ -88,20 +91,28 @@ def _prepare_for_class(self, inputs_dict, model_class, return_labels=False):
 
         if return_labels:
             if model_class in TF_MODEL_FOR_MULTIPLE_CHOICE_MAPPING.values():
-                inputs_dict["labels"] = tf.ones(self.model_tester.batch_size)
+                inputs_dict["labels"] = tf.ones(self.model_tester.batch_size, dtype=tf.int32)
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:94]: Like Patrick's comment above, a lot of the tests have identical code in their conditions.

File: tests/test_modeling_tf_common.py [hunk 6]
```diff
@@ -517,6 +528,47 @@ def test_determinism(self):
...
+            dict_inputs = self._prepare_for_class(inputs_dict, model_class, return_labels=True)
+            check_equivalence(model, tuple_inputs, dict_inputs)
+
+            tuple_inputs = self._prepare_for_class(inputs_dict, model_class)
+            dict_inputs = self._prepare_for_class(inputs_dict, model_class)
+            check_equivalence(model, tuple_inputs, dict_inputs, {"output_hidden_states": True})
+
+            tuple_inputs = self._prepare_for_class(inputs_dict, model_class)
+            dict_inputs = self._prepare_for_class(inputs_dict, model_class)
+            check_equivalence(model, tuple_inputs, dict_inputs, {"output_attentions": True})
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:570]: Can we add 4 more:
```
return_labels=True, output_hidden_states=True
return_labels=True, output_attentions=True
output_hidden_states=True, output_attentions=True
return_labels=True, output_hidden_states=True, output_hidden_states=True
```
for the sake of completeness?

**Review by sgugger — APPROVED** [review]
Looks great to me! Thanks for all the work!

**Comment by sgugger** [comment]
Side note, you should double-check the slow tests `test_saved_model_with_attentions_output` and `test_saved_model_with_hidden_states_output` still pass with the changes for the longformer model, as they are the ones that fail for t5 when we remove the `cast_to_bool`  thingy.

**Comment by patrickvonplaten** [comment]
> Side note, you should double-check the slow tests `test_saved_model_with_attentions_output` and `test_saved_model_with_hidden_states_output` still pass with the changes for the longformer model, as they are the ones that fail for t5 when we remove the `cast_to_bool` thingy.

They did not pass with Longformer before as discussed with @jplu on the PR: https://github.com/huggingface/transformers/pull/5764#issuecomment-670002430, they should actually pass now I think :-)

File: tests/test_modeling_tf_common.py (hunk 6 above)
Comment (LysandreJik) [reply tests/test_modeling_tf_common.py:570]: Yes, we can!
